app/src/util/pdf_utils.py

```diff
@@ -0,0 +1,60 @@
+from dataclasses import dataclass
+from typing import Any
+
+from pdfminer.pdfdocument import PDFDocument, PDFNoOutlines
+from pdfminer.pdfpage import PDFPage
+from pdfminer.pdftypes import PDFObjRef, resolve1
+from pdfminer.psparser import PSLiteral
+
+
+@dataclass
+class Heading:
+    title: str
+    # Heading level starting from 1
+    level: int
+    # Page number where the heading first appears
+    pageno: int | None
+
+
+def extract_outline(doc: PDFDocument) -> list[Heading]:
+    """
+    Adapted from dumppdf.py:dumpoutline()
+    Extracts the heading hierarchy from a PDF's catalog dictionary entry with key "Outlines".
+
+    Usage:
+        with open("100.pdf", "rb") as fp:
+            doc = PDFDocument(PDFParser(fp))
+            outline = extract_outline(doc)
```

Comment:
What do you think about changing the signature to take a filepointer and moving `doc = PDFDocument(PDFParser(fp))` inside the code here so this becomes `outline=extract_outline(fp)`? The benefit is that callers of this utility don't need to know anything about the internal functionality relying on pdfminer.six.

Reply:
Will do.